Add support for equatorial_radius and flattening in BaseGeodeticRepresentation.

cmarmo opened the pull request:
### Description
This pull request adds support for `equatorial_radius` and `flattening` in `BaseGeodeticRepresentation`.
~A new class `StandardGeodeticRepresentation` is added inheriting from `BaseGeodeticRepresentation`, retrieving~ ~spheroid attributes from a standard code.~
All geodetic representations classes are factorized into `geodetic.py`.

Tests are added.

Follows the suggestion in https://github.com/astropy/astropy/pull/14714#pullrequestreview-1407107055.

Towards https://github.com/astropy/astropy/issues/11170.

Thanks for considering it.

This work is funded by the Europlanet 2024 Research Infrastructure (RI) Grant.

cmarmo commented:
I am able to reproduce the test failure locally with
```
python -m tox -e py311-test-double
```
I am still trying to understand why ... :thinking:

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,107 @@
+
+@format_doc(geodetic_base_doc)
+class BaseGeodeticRepresentation(BaseRepresentation):
+    """Base geodetic representation."""
```
Maybe add a comment to make it explicit what subclasses need to do:
```
"""Base class for geodetic representations."""
# Subclasses need to set attributes _equatorial_radius and _flattening
# to quantities holding correct values (with units of length and dimensionless,
# respectively).
"""

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
 import pytest
-from numpy.testing import assert_array_equal
 
+from numpy.testing import assert_array_equal
```
I think the numpy line should stay where it is, above a blank line with astropy imports.

cmarmo replied:
Indeed, sorry, I forgot to revert after some experimenting...

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
 
+from numpy.testing import assert_array_equal
 from astropy import units as u
+from astropy.units import allclose as quantity_allclose
```
Are you using `pre-commit`? I think this resort will not survive it!

cmarmo replied:
Not sure I understand this comment.
The change was another mistake on my side: I've reverted and put the line where it was at the beginning.
I am indeed using pre-commit and it did not complain.
I have synchronized with upstream main (and rebased), just in case the pre-commit configuration had changed.

cmarmo replied:
...now that I'm fighting with ruff I understand the comment...

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
-from astropy.units import allclose as quantity_allclose
+from astropy.units.tests.test_quantity_erfa_ufuncs import vvd
+
+
```
Since you are defining new representations, you need to reset the `REPRESENTATION_CLASSES` dict after each test - otherwise they linger (I think this may be the reason that `test-double` failes). Probably easiest to just import `setup_function` and `teardown_function` from `test_representation`.

cmarmo replied:
It is not .... apparently pre-commit removes my import ... I guess I'm doing something wrong....

mhvk replied:
No, you do need to ensure that `REPRESENTATION_CLASSES` is set back to its original contents; your custom classes now cause test failures for `table` (I think, didn't look too carefully). I think you need
```
from astropy.coordinates.tests.test_representation import setup_function, teardown_function
````

cmarmo replied:
I did it! I didn't realize ruff deleted it ... I've added ```# ruff : noqa: 401``` but it does not force ruff to do what I want...

cmarmo replied:
I've edited `pyproject.toml` ... hope this is not a problem...

mhvk replied:
Sorry, I missed this chain - hope my suggested `# noqa: F401` will do the trick...

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
+        x=[1, 3000.0] * u.km, y=[7000.0, 4.0] * u.km, z=[5.0, 6000.0] * u.km
+    )
+
+    custom_geodetic = BaseGeodeticRepresentation
```
You should actually define a subclass, otherwise you are setting attributes on the real class which can come back to bite you. I.e.,
```
class CustomGeodetic(BaseGeodeticRepresentation):
    _flattening = 0.075 * u.dimensionless_unscaled
````

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
+    custom_geodetic = BaseGeodeticRepresentation
+    custom_geodetic._flattening = 0.075 * u.dimensionless_unscaled
+    with pytest.raises(AttributeError, match=msg):
+        s2 = custom_geodetic.from_representation(s1)
```
No need to assign to `s2` since this will fail anyway.

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
+        x=[1, 3000.0] * u.km, y=[7000.0, 4.0] * u.km, z=[5.0, 6000.0] * u.km
+    )
+
+    custom_geodetic = BaseGeodeticRepresentation
```
As above, do actually define a new subclass!

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
+
+    s3 = CartesianRepresentation.from_representation(s2)
+
+    s4 = custom_geodetic.from_representation(s3)
```
Why this extra `s4`? Certainly good to check that one can recover the cartesian representation, but I think a further round-trip test would make sense only if you started with a known geodetic location (maybe good to add?)

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,16 +2,65 @@
+    assert quantity_allclose(s2.lat, s4.lat)
+    assert quantity_allclose(s2.height, s4.height)
+
+    # Test initializer just for the sake of it.
```
So, I think I would *start* with this, but use explicit lon, lat, height.

cmarmo replied:
Answering to this comment and the previous one.
I have added round-trip tests from cartesian and from geodetic representation, each with its own initialization.

mhvk commented on docs/changes/coordinates/14763.feature.rst:
```diff
@@ -0,0 +1,4 @@
+Support for ``_equatorial_radius`` and ``_flattening`` is added in
```
I think we need a what's-new with an example of creating a geodetic subclass for some planet!

mhvk commented on docs/changes/coordinates/14763.feature.rst:
```diff
@@ -0,0 +1,4 @@
+Support for ``_equatorial_radius`` and ``_flattening`` is added in
+``BaseGeodeticRepresentation``.
+A new class ``StandardGeodeticRepresentation`` is added, inheriting from
```
Right now, at some level this is an implementation detail. On the other hand, if anyone ever subclasses `BaseRepresentation` themselves, code would no longer work - they now need to subclass something else.

I think best may be to use different names! See other in-line comments.

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,107 @@
+
+
+@format_doc(geodetic_base_doc)
+class BaseGeodeticRepresentation(BaseRepresentation):
```
I think we should have a more general name, since *geo*detic is for the Earth. How about `BasePlanetodeticRepresentation` (from [wikipedia](https://en.wikipedia.org/wiki/Planetary_coordinate_system)). 

Of course, this also begs the question of how the module should be called....

mhvk reviewed (commented):
Thanks, this looks great, and hence most of my comments (and perhaps confusion) are about names...  I haven't quite figured out the best way forward, but think I better just submit my present review since you may have better ideas.

p.s. Review is far easier with things split out - thanks very much for humouring me!

mhvk commented:
Having just added the "whatsnew-needed" label, I think what we also have to do is to add some documentation... I think a short additional section at the end of https://docs.astropy.org/en/latest/coordinates/representations.html#creating-your-own-representations should be good enough for now.

cmarmo commented:
@mhvk thanks for your review!
Before addressing your comments, I would like to clarify the issue about file and class names.

1. My understanding is that the representation module-to-be contains classes representing different ways of describing the coordinates. The current `BaseGeodeticClass` is defined by two angles and a length, where the latitude is defined by the angle between the x-axis and the vertical to the surface at a specific point of the spheroid.
Be the spheroid defined by a standard code or flattening and equatorial radius, does not change the essence of the representation. This is not the same as eg the spherical representation and the unit sphere representation: in that case the needed coordinates really change (r is not needed in the second).
That said, I would like to ask you to reconsider the creation of a new class (the one I called `StandardGeodeticRepresentation`): this will automatically solve some of your comments (https://github.com/astropy/astropy/pull/14763#discussion_r1186906057 and https://github.com/astropy/astropy/pull/14763#discussion_r1186909551) also forcing me to add checks about which attribute is defined like in #14714 (and answering https://github.com/astropy/astropy/pull/14763#discussion_r1186904519).
I understand the code itself looks cleaner here, but in my opinion an additional level of abstraction is unnecessary as it does not correspond to a different representation, and perhaps we are overcomplicating things for a user.

2. I believe we can safely use the name `geodetic.py` for the file containing the geodetic representation (even when generalized to other planets) as this is already done by the [NAIF SPICE routines](https://naif.jpl.nasa.gov/naif/index.html): for example the routines corresponding to `gc2gde` and `gd2gc` are [recgeo](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/cspice/recgeo_c.html) and [georec](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/cspice/georec_c.html).
A useful summary is available [here](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/Tutorials/pdf/individual_docs/17_frames_and_coordinate_systems.pdf#page=28): geodetic is the term used by NAIF SPICE "Applicable to bodies with equal equatorial radii."
We may be more precise using planetodetic but then we might want to change the names of all the functions defined in `EarthLocation` too and I'm wondering if this is worth it.

Let me know how do you feel about those suggestions.
Thanks!

cmarmo commented:
> Anyway, it sounds like we have a plan!

Wonderful! :)
I have addressed some comments ... if you don't mind I'm going to resolve some comments : I realized I'm starting to miss some inline comments and I need a way to check those I have addressed.

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,136 @@
+# Note: they get filled by the creation of the geodetic classes.
+
+
+def _check_ellipsoid(ellipsoid=None, default="WGS84"):
```
This is not used in this module, correct? If so, I think I'd just keep it in `earth.py`

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,136 @@
+class BaseGeodeticRepresentation(BaseRepresentation):
+    """
+    Base class for geodetic representations.
+    Subclasses need to set attributes ``_equatorial_radius`` and ``_flattening``
```
Insert an empty line here so that the docstring gets typeset properly. See https://astropy--14763.org.readthedocs.build/en/14763/api/astropy.coordinates.BaseGeodeticRepresentation.html#astropy.coordinates.BaseGeodeticRepresentation

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,136 @@
+    Subclasses need to set attributes ``_equatorial_radius`` and ``_flattening``
+    to quantities holding correct values (with units of length and dimensionless,
+    respectively), or alternatively an ``_ellipsoid`` attribute to the relevant ERFA
+    index (as passed in to erfa.eform).
```
It should work to put back ticks around `erfa.eform` so one gets a link to its documentation. (I.e., links to https://pyerfa.readthedocs.io/en/latest/api/erfa.eform.html#erfa.eform)

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,136 @@
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        if "_ellipsoid" in cls.__dict__:
+            ELLIPSOIDS[cls._ellipsoid] = cls
```
Do this at the end of the stanza, so that if `_ellipsoid` is not a proper entry, we get an error message without having added an entry to `ELLIPSOIDS` (indeed, might be worth testing that!)

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,136 @@
+            or "_flattening" not in cls.__dict__
+        ):
+            raise AttributeError(
+                f"{cls.__name__} requires `_ellipsoid` or _equatorial_radius and _flattening."
```
Total nitpick, but I'd write
```
f"(cls.__name__} should define either '_ellipsoid' or '_equatorial_radius' and '_flattening'."
```

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,49 +2,97 @@
+    with pytest.raises(AttributeError, match=msg):
+
+        class CustomGeodeticError(BaseGeodeticRepresentation):
+            _flattening = 0.075 * u.dimensionless_unscaled
```
Here, you could add
```
assert "customgeodeticerror" not in REPRESENTATION_CLASSES
```

mhvk commented on astropy/coordinates/geodetic.py:
```diff
@@ -0,0 +1,136 @@
+    attr_classes = {"lon": Longitude, "lat": Latitude, "height": u.Quantity}
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
```
Put the `super` call at the end, so that it gets executed only when the part here is successful, otherwise you register representations even if they do not initialize properly.

mhvk reviewed (commented):
Now definitely moving towards minimal comments and nitpicks - I think overall this is looking very good!

But obviously need to figure out why the tests are failing. My guess is that it is the registration of the test classes, and I give a suggestion in-line on how to fix this...

To do is a what's-new entry and documentation!

cmarmo commented on docs/coordinates/representations.rst:
```diff
@@ -685,3 +693,22 @@ In pseudo-code, this means that a class will look like::
+    class IAUMARS1979GeodeticRepresentation(BaseGeodeticRepresentation):
+
+        _equatorial_radius = 3393400.0 * u.m
+        _flattening = 0.518650e-2 * u.dimensionless_unscaled
```
Having to add `* u.dimensionless_unscaled` to attribute is not really user friendly in my opinion.
What about a check on `_flattening` adding the units when it is missing?

cmarmo replied:
It was a bug then.
I have modified the tests and the code, otherwise when a float was passed the conversion failed.

mhvk commented on docs/whatsnew/6.0.rst:
```diff
@@ -22,6 +22,29 @@ By the numbers:
+    >>> representation.to_cartesian()
+    <CartesianRepresentation (x, y, z) in m
+        (2481113.87541373, 2554650.51055863, 5274062.32189811)>
+
```
Would it make sense to also add `representation.represent_as(WGS84GeodeticRepresentation)`? It helps the inquisitive user understand there are already some predefined classes.

mhvk commented on docs/whatsnew/6.0.rst:
```diff
@@ -22,6 +22,29 @@ By the numbers:
+    ...     _equatorial_radius = 6378140 * u.m
+    ...     _flattening = 0.003352805 * u.dimensionless_unscaled
+    >>> representation = IAU1976EarthGeodeticRepresentation(lon=0.8*u.radian,
+    ...     lat=0.98*u.radian, height=367*u.m)
```
This is really personal preference, but I think one might as well use `u.deg` for the example, since that is a more natural unit. Or, if one were to be fancy, something like `'32d10mW', '10d20mN'`

mhvk commented on pyproject.toml:
```diff
@@ -309,6 +309,7 @@ unfixable = [
 "setup.py" = ["INP001"]  # Part of configuration, not a package.
 "__init__.py" = ["E402", "F401", "F403"]
 "astropy/coordinates/angle_formats.py" = ["D415"]  # The docstrings are regexes.
+"astropy/coordinates/tests/test_geodetic_representations.py" = ["F401"]  # Unused import needed.
```
Good that you found that! Though I think in this case a `# noqa: F401` at the import line is perhaps clearer... These lines really are more for cases were the code would otherwise just be littered with `noqa`.

mhvk commented on docs/whatsnew/6.0.rst:
```diff
@@ -22,6 +22,29 @@ By the numbers:
+Define Geodetic Representations via their geometric parameters
+==============================================================
+
+The `~astropy.coordinates.BaseGeodeticRepresentation` now exposes its geometric
```
I think this first sentence is a bit of an "implementation detail" that users do not really have to care about. Maybe:
```
The user may now define custom spheroidal models for the Earth or other planetary
bodies by subclassing `~astropy.coordinates.BaseGeodeticRepresentation` and defining
``_equatorial_radius`` and ``_flattening`` attributes::

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,49 +2,116 @@
+
+# Preserve the original REPRESENTATION_CLASSES dict so that importing
+#   the test file doesn't add a persistent test subclass (LogDRepresentation)
+from astropy.coordinates.tests.test_representation import (
```
Here, add `# noqa: F401` instead of the extra line in `pyproject.toml`.

cmarmo replied:
I had to commit with `--no-verify` in order to make this syntax pass.
And the pre-commit check here is red.

mhvk reviewed (commented):
Great! Code is now all OK - the diff makes me feel this is exactly right, with just a clean chunk moved from `earth.py` to the new `geodetic.py`, and logical additions.

Most of the remaining comments are just textual, on the added documentation.

One thought (not for this PR!!): it might make sense to allow one to input a `BaseGeodeticRepresentation` subclass in `EarthLocation.to_geodetic` (maybe as the `ellipsoid` argument), so people can use their custom classes in there.

eerovaher commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,49 +2,116 @@
+    msg = "module 'erfa' has no attribute 'foo'"
+    with pytest.raises(AttributeError, match=msg):
 
+        class CustomGeodeticEllipsoidError(BaseGeodeticRepresentation):
```
If I read the code I can figure out why an error is implemented as a subclass of `BaseGeodeticRepresentation`, but I'd rather not spend the effort.
```suggestion
        class InvalidCustomGeodeticEllipsoid(BaseGeodeticRepresentation):
```
This also applies to the other class in this function.

eerovaher commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,49 +2,116 @@
-    assert quantity_allclose(s1.y, s3.y)
-    assert quantity_allclose(s1.z, s3.z)
+    # Test geodetic initialization round-trip
+    s1g = CustomGeodetic(
+        lon=[0.8, 1.3] * u.radian,
+        lat=[0.3, 0.98] * u.radian,
+        height=[100.0, 367.0] * u.m,
+    )
+
+    s2g = CartesianRepresentation.from_representation(s1g)
+
+    s3g = CustomGeodetic.from_representation(s2g)
```
I don't like having variables that are used exactly once because I don't know that they'll be discarded almost immediately when I'm reading the code, so I try to memorize what they are only to find out very quickly that I can forget about them entirely.
```suggestion
    initial = CustomGeodetic(
        lon=[0.8, 1.3] * u.radian,
        lat=[0.3, 0.98] * u.radian,
        height=[100.0, 367.0] * u.m,
    )
    roundtripped = CustomGeodetic.from_representation(
        CartesianRepresentation.from_representation(initial)
    )
```

eerovaher commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,49 +2,116 @@
+    assert "customgeodeticerror" not in REPRESENTATION_CLASSES
+
+
+def test_custombasegeodetic_roundtrip():
```
This test should be [parametrized](https://docs.pytest.org/en/7.3.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions). If one of the asserts fails then the following asserts in the test function are not executed, but in the case of a parametrized test one of the parametrizations failing does not prevent other parametrizations of the same test from running, so a parametrized test suite is more granular and informative with no or very little extra effort.

eerovaher commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -2,49 +2,116 @@
+
+    s3g = CustomGeodetic.from_representation(s2g)
+
+    assert quantity_allclose(s1g.lon, s3g.lon)
```
It is better to use [`assert_quantity_allclose()`](https://docs.astropy.org/en/v5.2.2/api/astropy.tests.helper.assert_quantity_allclose.html) because it is more informative if the assertion fails.

eerovaher reviewed (commented):
I have some comments regarding the tests.

cmarmo commented on docs/coordinates/representations.rst:
```diff
@@ -30,6 +30,14 @@ The built-in representation classes are:
 * `~astropy.coordinates.CylindricalRepresentation`:
   cylindrical polar coordinates, represented by a cylindrical radius
   (``rho``), azimuthal angle (``phi``), and height (``z``).
+* `~astropy.coordinates.BaseGeodeticRepresentation`:
```
> This class is different in that it is not useful in itself but has to be subclassed.

I understand this point.
But then `WGS84GeodeticRepresentation` is a specific case, which is not true for the other items in the list.
What about removing the geodetic description from the list and adding a paragraph like:
"Astropy also offers a BaseGeodeticRepresentation useful to create specific representations on spheroidal bodies... blah...blah..."

cmarmo commented:
@eerovaher , Thanks for your comments, I believe I have addressed all your concern?

@mhvk something is wrong with the conversion of `_flattening`... I am confused.
I had to manually force the conversion between a float and a `dimensionless_unscaled`, also the docstring is failing with a `%` since I have defined the flattening as `u.percent` ...

mhvk commented:
@cmarmo - the reason that `_flattening` didn't work is that we didn't quite do the `ufunc` right for that - my mistake to not notice that, really. I made a change locally, and also some small further changes to the tests (to split them up further as requested by @eerovaher and minimize the difference with what was there), but I cannot push to your branch. Could you allow me to? (Need to "Enable edits by maintainers".). If that's tricky, the rebased/squashed branch including a commit for the helpers is at https://github.com/mhvk/astropy/tree/representations-geodetic - you could get that and push to your branch yourself.

mhvk commented:
p.s. Just updated my branch with an additional commit that moves `geodetic` inside `astropy.coordinates.representations` (which was made its own submodule in #14792). It also corrected your commit to put back in that infamous `noqa: F401` which also for me had been automatically removed.

cmarmo commented:
I cannot check the "allow edits from maintainers" because of https://github.com/orgs/community/discussions/5634.
I'm going to make some manipulations locally ... back soon.

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -31,6 +31,18 @@ The built-in representation classes are:
 
+
+Astropy also offers a `~astropy.coordinates.BaseGeodeticRepresentation` useful to
+create specific representations on spheroidal bodies (see
```
I think this paragraph is very clear! Should we mention the existing earth-based ones? E.g.,
```

Astropy also offers a `~astropy.coordinates.BaseGeodeticRepresentation` useful to
create specific `repreentations` on spheroidal bodies. This is used internally for the
standard Earth ellipsoids used in `~astropy.coordinates.EarthLocation` 
(`~astropy.coordinates.WGS84GeodeticRepresentation`,
`~astropy.coordinates.WGS72GeodeticRepresentation`, and
`~astropy.coordinates.GRS80GeodeticRepresentation`), but 
can also be customized; see :ref:`astropy-coordinates-create-geodetic`.
All these are coordinates on ...
```

cmarmo replied:
I like it! I'm adding it right now.

mhvk reviewed (approved):
Sorry for the trouble, I hadn't realized group accounts were different!
And also sorry that my PR missed some things!

But it now all looks great! I have one possible suggestion for the documentation, but am also happy to merge as is, so will leave that up to you and approve now.

mhvk commented:
The tests seem to be stuck, so I'm going to close & reopen. Hopefully, that will resolve it. Otherwise, everything is OK, so I'm going to try our new "auto-merge" feature

cmarmo commented:
Thanks! :partying_face:
